include/pybind11/detail/function_view.h

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2021 The Pybind Development Team.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "common.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename Signature>
+class function_view;
+
+template <typename R, typename... Arg>
+class function_view<R(Arg...)> {
+public:
+    function_view() = default;
+
+    /// Creates a view that references an existing function.
+    ///
+    /// The constructed view is only valid for the lifetime of `f`.
+    ///
+    /// Requires that `F` is a function object type that is callable with
+    /// arguments `(Args...)` and has a return type convertible to `R`.
+    template <typename F,
+              typename = enable_if_t<
+                  // Prevent construction from the same `function_view` type, to avoid inefficient
+                  // double wrapping that also commonly leads to dangling references.
+                  (!std::is_same<remove_cvref_t<F>, function_view>::value
+                   // Ensure `F` is compatible with the signature.
+                   && is_expected_return_type<R>::template compatible_with<
+                       call_result_t<F, Arg...>>::value)>>
+    function_view(F &&f) noexcept // NOLINT(google-explicit-constructor)
+        : erased_fn_(&wrapper<remove_reference_t<F>>), erased_obj_(&f) {}
+
+    /// Calls the referenced function.
+    ///
+    /// Precondition: this is a non-null function view.
+    R operator()(Arg... arg) const {
+        return erased_fn_(const_cast<void *>(erased_obj_), static_cast<Arg &&>(arg)...);
+    }
+
+    /// Returns `true` if this is a non-null function view.
+    explicit operator bool() const { return erased_fn_ != nullptr; }
+
+private:
+    template <typename F>
+    static R wrapper(void *obj, Arg... arg) {
+        return (*static_cast<typename std::add_pointer<F>::type>(obj))(
+            static_cast<Arg &&>(arg)...);
+    }
+
+    R (*erased_fn_)(void *, Arg...) = nullptr;
+    const void *erased_obj_;
```

Comment:
`type_erased_obj_`?
If that's correct, I'd definitely spend the extra few characters for clarity.